<a href="https://colab.research.google.com/github/cameronabel/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/CAMERON_ABEL_LS_DS_112.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200> 

# Assignment:

- Replicate the lesson code.

 - This means that if you haven't followed along already, type out the things that we did in class. Forcing your fingers to hit each key will help you internalize the syntax of what we're doing. Make sure you understand each line of code that you're writing, google things that you don't fully understand.
 - [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)
- Convert the `term` column from string to integer.
- Make a column named `loan_status_is_great`. It should contain the integer 1 if `loan_status` is "Current" or "Fully Paid." Else it should contain the integer 0.
- Make `last_pymnt_d_month` and `last_pymnt_d_year` columns.

In [0]:
##### Begin Working Here #####

In [0]:
import pandas as pd

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')
df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

In [4]:
pd.set_option('display.max_rows', 100)
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley             object
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure      object
BsmtFinType1      object
BsmtFinSF1         int64
BsmtFinType2      object
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating           object


In [5]:
# Though I prefer the following for examining what datatypes may be present in the data
df.dtypes.value_counts()

object     43
int64      35
float64     3
dtype: int64

In [6]:
# Look at the first ten rows of the `BedroomAbvGr` column.

df[['BedroomAbvGr']].head(10)

BedroomAbvGr
0             3
1             3
2             3
3             3
4             4
5             1
6             3
7             3
8             2
9             2

In [8]:
# Or via dot notation
df.BedroomAbvGr.head(10)

0    3
1    3
2    3
3    3
4    4
5    1
6    3
7    3
8    2
9    2
Name: BedroomAbvGr, dtype: int64

In [9]:
# Look at the first ten rows of the `LotFrontage` column
df['LotFrontage'].head(10)

0    65.0
1    80.0
2    68.0
3    60.0
4    84.0
5    85.0
6    75.0
7     NaN
8    51.0
9    50.0
Name: LotFrontage, dtype: float64

All values look like they should be ints, but the NaN is throwing us into float territory.

In [10]:
df['LotFrontage'].value_counts(dropna=False)

NaN      259
60.0     143
70.0      70
80.0      69
50.0      57
        ... 
106.0      1
38.0       1
138.0      1
140.0      1
137.0      1
Name: LotFrontage, Length: 111, dtype: int64

In [0]:
import numpy as np

In [12]:
type(np.nan)

float

Confirming that np.nan is float.

###Examining a few specific columns:

In [15]:
small_df = df[['TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'SalePrice']].copy()
small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice
0          856       856       854     208500
1         1262      1262         0     181500
2          920       920       866     223500
3          756       961       756     140000
4         1145      1145      1053     250000

####Adding a total square footage column:

In [16]:
small_df['TotalSF'] = small_df['TotalBsmtSF'] + small_df['1stFlrSF'] + small_df['2ndFlrSF']
small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSF
0          856       856       854     208500     2566
1         1262      1262         0     181500     2524
2          920       920       866     223500     2706
3          756       961       756     140000     2473
4         1145      1145      1053     250000     3343

And calculating price per SF:

In [17]:
small_df['PricePerSF'] = small_df['SalePrice'] / small_df['TotalSF']
small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSF  PricePerSF
0          856       856       854     208500     2566   81.254871
1         1262      1262         0     181500     2524   71.909667
2          920       920       866     223500     2706   82.594235
3          756       961       756     140000     2473   56.611403
4         1145      1145      1053     250000     3343   74.783129

## __Moving on to the lending dataset__

In [18]:
! wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-01-18 17:30:47--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 64.48.1.20
Connecting to resources.lendingclub.com (resources.lendingclub.com)|64.48.1.20|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [                 <=>]  21.72M   914KB/s    in 25s     

2020-01-18 17:31:12 (897 KB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [22776791]



In [19]:
! unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


In [20]:
# Using bash to examine the head

! head LoanStats_2018Q4.csv 

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [21]:
# Using bash to examine the tail

! tail LoanStats_2018Q4.csv 

"","","5600","5600","5600"," 36 months"," 13.56%","190.21","C","C1","","n/a","RENT","15600","Not Verified","Oct-2018","Current","n","","","credit_card","Credit card refinancing","836xx","ID","15.31","0","Aug-2012","0","","97","9","1","5996","34.5%","11","w","3686.37","3686.37","2652.39","2652.39","1913.63","738.76","0.0","0.0","0.0","Dec-2019","190.21","Jan-2020","Nov-2019","0","","1","Individual","","","","0","0","5996","0","0","0","1","20","0","","0","2","3017","35","17400","1","0","0","3","750","4689","45.5","0","0","20","73","13","13","0","13","","20","","0","3","5","4","4","1","9","10","5","9","0","0","0","0","100","25","1","0","17400","5996","8600","0","","","","","","","","","","","","N","","","","","","","","","","","","","","","N","","","","","",""
"","","23000","23000","23000"," 36 months"," 15.02%","797.53","C","C3","Tax Consultant","10+ years","MORTGAGE","75000","Source Verified","Oct-2018","Charged Off","n","","","debt_consolidation","Debt consolidation","352xx","AL","20.9

Switching to Python:

In [0]:
df = pd.read_csv('LoanStats_2018Q4.csv', header=1, skipfooter=2, engine='python')

In [35]:
df.shape

(128397, 144)

In [30]:
# Examining null values

pd.set_option('display.max_rows', 200)
df.isnull().sum().sort_values()

inq_fi                                             0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
inq_last_12m                                       0
total_cu_tl                                        0
total_rev_hi_lim                                   0
max_bal_bc                                         0
open_rv_24m                                        0
open_rv_12m                                        0
total_bal_il                                       0
open_il_24m                                        0
hardship_flag                                      0
open_act_il                                        0
open_acc_6m                                        0
tot_cur_bal                                        0
tot_coll_amt                                       0
acc_now_delinq                                     0
application_type                              

##Lecture question: Why might LendingClub have included columns in their dataset that are 100% blank?

The null values in the dataset look like they are identifiers for LendingClub to use internally and are not meant to be published.

In [0]:
# Dropping columns that are full of nulls

df = df.drop(['url', 'member_id', 'desc', 'id'], axis=1)

###Cleaning up the interest rates:

In [37]:
df['int_rate'].head(10)

0     14.47%
1     10.33%
2     15.02%
3     12.98%
4     16.14%
5     12.98%
6     15.02%
7     16.14%
8     14.47%
9     13.56%
Name: int_rate, dtype: object

In [38]:
df['int_rate'][3]

' 12.98%'

In [39]:
type(df['int_rate'][3])

str

We can see that the interest rate column is composed of strings.  The string have a leading space and a trailing %, with the relevant numerical data sandwiched in-between.  We need to parse these fields to remove both characters and convert to float.

For something this simple, I would probably just just a lambda function.  Per the lecture, we can build a new series first and then add that series to the df

Via lambda function:

In [0]:
df['int_rate_cleaned'] = df['int_rate'].apply(lambda cell: float(cell.strip().strip('%')))

Via new series:

In [0]:
def myfunc(cell_contents):
  return float(cell_contents.strip().strip('%'))

In [0]:
new_list = [myfunc(cell) for cell in df['int_rate']]

In [47]:
df['int_rate_cleaned'] = pd.Series(new_list)
df.head()

loan_amnt  funded_amnt  ...  settlement_term int_rate_cleaned
0       6000         6000  ...             24.0            14.47
1      24000        24000  ...              NaN            10.33
2      30000        30000  ...              NaN            15.02
3       8000         8000  ...              NaN            12.98
4      22400        22400  ...              NaN            16.14

[5 rows x 141 columns]

In [51]:
#confirming type is now float

df['int_rate_cleaned'].dtypes

dtype('float64')

Cleaning up employment titles:

In [0]:
df['emp_title'] = df['emp_title'].fillna('Unknown').apply(lambda cell: cell.title().strip())

In [57]:
df['emp_title']

0                         Manager
1             Maintenance Manager
2                     Firefighter
3                      Controller
4         Paramedic/ Fire Fighter
                   ...           
128392             Tax Consultant
128393             Security Guard
128394               Payoff Clerk
128395                    Unknown
128396       Key Accounts Manager
Name: emp_title, Length: 128397, dtype: object

In [59]:
# Checking to ensure all null values were overwritten
df['emp_title'].isnull().sum()

0

In [60]:
# Checking unique values
len(df['emp_title'].unique())

34899

##Handling datetimes

In [64]:
# Identifying relevant columns:

date_cols = [col for col in df if col.endswith('_d') or col.endswith('_date')]
date_cols

['issue_d',
 'last_pymnt_d',
 'next_pymnt_d',
 'last_credit_pull_d',
 'hardship_start_date',
 'hardship_end_date',
 'payment_plan_start_date',
 'debt_settlement_flag_date',
 'settlement_date']

In [62]:
df['issue_d'][:10]

0    Dec-2018
1    Dec-2018
2    Dec-2018
3    Dec-2018
4    Dec-2018
5    Dec-2018
6    Dec-2018
7    Dec-2018
8    Dec-2018
9    Dec-2018
Name: issue_d, dtype: object

Converting to datetime objects:

In [65]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)
df[date_cols].head()

issue_d last_pymnt_d  ... debt_settlement_flag_date settlement_date
0 2018-12-01     Apr-2019  ...                  Oct-2019        Oct-2019
1 2018-12-01     Dec-2019  ...                       NaN             NaN
2 2018-12-01     Feb-2019  ...                       NaN             NaN
3 2018-12-01     Dec-2019  ...                       NaN             NaN
4 2018-12-01     Dec-2019  ...                       NaN             NaN

[5 rows x 9 columns]

In [66]:
# Examining the contents of one cell

df['issue_d'].iloc[0]

Timestamp('2018-12-01 00:00:00')

In [68]:
# Now to examine an array of the first 5 cells:

df['issue_d'].head().values

array(['2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000'], dtype='datetime64[ns]')

###Building a new DataFrame just to contain date information:

In [69]:
date_df = df[date_cols].copy()
date_df['year'] = date_df['issue_d'].dt.year
date_df.head()

issue_d last_pymnt_d  ... settlement_date  year
0 2018-12-01     Apr-2019  ...        Oct-2019  2018
1 2018-12-01     Dec-2019  ...             NaN  2018
2 2018-12-01     Feb-2019  ...             NaN  2018
3 2018-12-01     Dec-2019  ...             NaN  2018
4 2018-12-01     Dec-2019  ...             NaN  2018

[5 rows x 10 columns]

In [70]:
date_df['month'] = date_df['issue_d'].dt.month
date_df.head()

issue_d last_pymnt_d next_pymnt_d  ... settlement_date  year month
0 2018-12-01     Apr-2019          NaN  ...        Oct-2019  2018    12
1 2018-12-01     Dec-2019     Jan-2020  ...             NaN  2018    12
2 2018-12-01     Feb-2019          NaN  ...             NaN  2018    12
3 2018-12-01     Dec-2019     Jan-2020  ...             NaN  2018    12
4 2018-12-01     Dec-2019     Jan-2020  ...             NaN  2018    12

[5 rows x 11 columns]

Exmining the types of the newly added 'year' and 'month' columns:

In [72]:
date_df.dtypes[-2:]

year     int64
month    int64
dtype: object

Examining the last credit pull dates:

In [74]:
date_df['last_credit_pull_d'] = pd.to_datetime(date_df['last_credit_pull_d'], infer_datetime_format=True)
date_df.head()

issue_d last_pymnt_d next_pymnt_d  ... settlement_date  year month
0 2018-12-01     Apr-2019          NaN  ...        Oct-2019  2018    12
1 2018-12-01     Dec-2019     Jan-2020  ...             NaN  2018    12
2 2018-12-01     Feb-2019          NaN  ...             NaN  2018    12
3 2018-12-01     Dec-2019     Jan-2020  ...             NaN  2018    12
4 2018-12-01     Dec-2019     Jan-2020  ...             NaN  2018    12

[5 rows x 11 columns]

Building a new column to show the length of credit history

In [75]:
date_df['last_credit_pull_d'] - date_df['issue_d']

0        335 days
1        335 days
2        335 days
3        365 days
4        335 days
           ...   
128392    31 days
128393   396 days
128394   396 days
128395   396 days
128396   396 days
Length: 128397, dtype: timedelta64[ns]

In [76]:
date_df['day_diff'] = (date_df['last_credit_pull_d'] - date_df['issue_d']).dt.days
date_df.head()

issue_d last_pymnt_d next_pymnt_d  ...  year month day_diff
0 2018-12-01     Apr-2019          NaN  ...  2018    12    335.0
1 2018-12-01     Dec-2019     Jan-2020  ...  2018    12    335.0
2 2018-12-01     Feb-2019          NaN  ...  2018    12    335.0
3 2018-12-01     Dec-2019     Jan-2020  ...  2018    12    365.0
4 2018-12-01     Dec-2019     Jan-2020  ...  2018    12    335.0

[5 rows x 12 columns]

###Lecture question: What's oldest credit history that was involved in Q4 2018?

This question relies on some dates that are not present in the date_df.

In [80]:
df.head()

loan_amnt  funded_amnt  ...  settlement_term int_rate_cleaned
0       6000         6000  ...             24.0            14.47
1      24000        24000  ...              NaN            10.33
2      30000        30000  ...              NaN            15.02
3       8000         8000  ...              NaN            12.98
4      22400        22400  ...              NaN            16.14

[5 rows x 141 columns]

I'm thinking I need to access the earliest_credit_line column in the original df:

In [81]:
df['earliest_cr_line'].sort_values()

105555    Apr-1960
5253      Apr-1962
12387     Apr-1963
19825     Apr-1963
33749     Apr-1963
            ...   
107305    Sep-2015
5284      Sep-2015
71118     Sep-2015
106301    Sep-2015
69416     Sep-2015
Name: earliest_cr_line, Length: 128397, dtype: object

The earliest credit line was issued in April of 1960.

In [95]:
pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True).min() - pd.to_datetime('12-1-2018')

Timedelta('-25171 days +00:00:00')

Timedelta is 25,171 days and ties to lecture notes.

##Additional assignments:
Convert the term column from string to integer.

Make a column named loan_status_is_great. It should contain the integer 1 if loan_status is "Current" or "Fully Paid." Else it should contain the integer 0.

Make last_pymnt_d_month and last_pymnt_d_year columns.

###Convert the term column from string to integer.

In [96]:
df.term.head()

0     36 months
1     36 months
2     36 months
3     36 months
4     60 months
Name: term, dtype: object

In [99]:
set([' months' in cell for cell in df.term])

{True}

We can see that every single cell in the term column contains ' months'

In [101]:
set([cell for cell in df.term])

{' 36 months', ' 60 months'}

Here we can see that there are only two possible values for each cell in the column.  Each has a leading space and a trailing ' months'

In [0]:
df.term = df.term.apply(lambda cell: int(cell[1:3]))

In [107]:
df['term'].head()

0    36
1    36
2    36
3    36
4    60
Name: term, dtype: int64

###Make a column named loan_status_is_great. It should contain the integer 1 if loan_status is "Current" or "Fully Paid." Else it should contain the integer 0.

In [0]:
df['loan_status_is_great'] = df['loan_status'].apply(lambda status: 1 if status in ['Current', 'Fully Paid'] else 0)

In [138]:
df[['loan_status', 'loan_status_is_great']].sample(10)

loan_status  loan_status_is_great
90952              Current                     1
38807              Current                     1
50910              Current                     1
71132              Current                     1
15100              Current                     1
91936              Current                     1
102665          Fully Paid                     1
6003               Current                     1
94728              Current                     1
56643   Late (31-120 days)                     0

###Make last_pymnt_d_month and last_pymnt_d_year columns.

In [0]:
df['last_pymnt_d_month'] = pd.to_datetime(df['last_pymnt_d'], infer_datetime_format=True).dt.month

In [0]:
df['last_pymnt_d_year'] = pd.to_datetime(df['last_pymnt_d'], infer_datetime_format=True).dt.year

In [142]:
df[['last_pymnt_d', 'last_pymnt_d_month', 'last_pymnt_d_year']].head()

last_pymnt_d  last_pymnt_d_month  last_pymnt_d_year
0     Apr-2019                 4.0             2019.0
1     Dec-2019                12.0             2019.0
2     Feb-2019                 2.0             2019.0
3     Dec-2019                12.0             2019.0
4     Dec-2019                12.0             2019.0

# Stretch Goals

You can do more with the LendingClub or Instacart datasets.

LendingClub options:
- There's one other column in the dataframe with percent signs. Remove them and convert to floats. You'll need to handle missing values.
- Modify the `emp_title` column to replace titles with 'Other' if the title is not in the top 20. 
- Take initiatve and work on your own ideas!

Instacart options:
- Read [Instacart Market Basket Analysis, Winner's Interview: 2nd place, Kazuki Onodera](http://blog.kaggle.com/2017/09/21/instacart-market-basket-analysis-winners-interview-2nd-place-kazuki-onodera/), especially the **Feature Engineering** section. (Can you choose one feature from his bulleted lists, and try to engineer it with pandas code?)
- Read and replicate parts of [Simple Exploration Notebook - Instacart](https://www.kaggle.com/sudalairajkumar/simple-exploration-notebook-instacart). (It's the Python Notebook with the most upvotes for this Kaggle competition.)
- Take initiative and work on your own ideas!

###There's one other column in the dataframe with percent signs. Remove them and convert to floats. You'll need to handle missing values.

In [143]:
df.head()

loan_amnt  funded_amnt  ...  last_pymnt_d_month  last_pymnt_d_year
0       6000         6000  ...                 4.0             2019.0
1      24000        24000  ...                12.0             2019.0
2      30000        30000  ...                 2.0             2019.0
3       8000         8000  ...                12.0             2019.0
4      22400        22400  ...                12.0             2019.0

[5 rows x 144 columns]

The offending column is 'revol_util'

In [144]:
df['revol_util'][0]

'91.2%'

In [0]:
df['revol_util_cleaned'] = df['revol_util'].apply(lambda cell: float(str(cell).strip('%')))

I found that I needed to cast the cell contents to str to handle the NaN values (was not necessary for the interest rates column).  Recasting to float converts the missing values back to np.nan.

###Modify the emp_title column to replace titles with 'Other' if the title is not in the top 20.

You can uncomment and run the cells below to re-download and extract the Instacart data

In [0]:
# !wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

In [0]:
# !tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

In [0]:
# %cd instacart_2017_05_01